book/src/01-limitations/03-volatile_destination.md

```diff
@@ -1 +1,345 @@
+friend, that compiler.
+
+Volatile memory works almost the opposite way. With volatile memory we
+need the compiler to _definitely_ emit an instruction to do a read or write and
+they need to happen _exactly_ in the order that we say to do it. Each volatile
+read or write might have any sort of side effect that the compiler
+doesn't know about, and it shouldn't try to be clever about the optimization. Just do what we
+say, please.
+
+In Rust, we don't mark volatile things as being a separate type of thing,
+instead we use normal raw pointers and then call the
+[read_volatile](https://doc.rust-lang.org/core/ptr/fn.read_volatile.html) and
+[write_volatile](https://doc.rust-lang.org/core/ptr/fn.write_volatile.html)
+functions (also available as methods, if you like), which then delegate to the
+LLVM
+[volatile_load](https://doc.rust-lang.org/core/intrinsics/fn.volatile_load.html)
+and
+[volatile_store](https://doc.rust-lang.org/core/intrinsics/fn.volatile_store.html)
+intrinsics. In C and C++ you can tag a pointer as being volatile and then any
+normal read and write with it becomes the volatile version, but in Rust we have
+to remember to use the correct alternate function instead.
+
+I'm told by the experts that this makes for a cleaner and saner design from a
+_language design_ perspective, but it really kinda screws us when doing low
+level code. References, both mutable and shared, aren't volatile, so they
+compile into normal reads and writes. This means we can't do anything we'd
+normally do in Rust that utilizes references of any kind. Volatile blocks of
+memory can't use normal `.iter()` or `.iter_mut()` based iteration (which give
+`&T` or `&mut T`), and they also can't use normal `Index` and `IndexMut` sugar
+like `a + x[i]` or `x[i] = 7`.
+
+Unlike with normal raw pointers, this pain point never goes away. There's no way
+to abstract over the difference with Rust as it exists now, you'd need to
+actually adjust the core language by adding an additional pointer type (`*vol
+T`) and possibly a reference type to go with it (`&vol T`) to get the right
+semantics. And then you'd need an `IndexVol` trait, and you'd need
+`.iter_vol()`, and so on for every other little thing. It would be a lot of
+work, and the Rust developers just aren't interested in doing all that for such
+a limited portion of their user population. We'll just have to deal with not
+having any syntax sugar.
+
+### VolatilePtr
+
+No syntax sugar doesn't mean we can't at least make things a little easier for
+ourselves. Enter the `VolatilePtr<T>` type, which is a newtype over a `*mut T`.
+One of those "manual" newtypes I mentioned where we can't use our nice macro.
+
+```rust
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord)]
+#[repr(transparent)]
+pub struct VolatilePtr<T>(pub *mut T);
+```
+
+Obviously we want to be able to read and write:
+
+```rust
+impl<T> VolatilePtr<T> {
+  /// Performs a `read_volatile`.
+  pub unsafe fn read(self) -> T {
```

Comment:
How about a smart pointer? `impl Deref` and `DerefMut` instead of adding methods?

Reply:
Oh, no no, we cannot do that, because references are never volatile, so reads and writes with them are normal reads and writes. I'll actually be sure to include that in the revision of this page (which will be soon but I need a PR to go into nightly first)